core/src/main/java/org/infinispan/persistence/manager/TracedPersistenceManager.java

```diff
@@ -0,0 +1,117 @@
+      return decorateCommand("loadFromAllStores", super.loadFromAllStores(key, segment, localInvocation, includeStores));
+   }
+
+   @Override
+   public CompletionStage<Long> approximateSize(Predicate<? super StoreConfiguration> predicate, IntSet segments) {
+      return decorateCommand("approximateSize", super.approximateSize(predicate, segments));
+   }
+
+   @Override
+   public CompletionStage<Long> size(Predicate<? super StoreConfiguration> predicate, IntSet segments) {
+      return decorateCommand("size", super.size(predicate, segments));
+   }
+
+   @Override
+   public CompletionStage<Void> writeToAllNonTxStores(MarshallableEntry marshalledEntry, int segment,
+                                                      Predicate<? super StoreConfiguration> predicate, long flags) {
+      return decorateCommand("writeToAllNonTxStores", super.writeToAllNonTxStores(marshalledEntry, segment, predicate, flags));
+   }
+
+   @Override
+   public CompletionStage<Void> commitAllTxStores(TxInvocationContext<AbstractCacheTransaction> txInvocationContext,
+                                                  Predicate<? super StoreConfiguration> predicate) {
+      return decorateCommand("commitAllTxStores", super.commitAllTxStores(txInvocationContext, predicate));
+   }
+
+   @Override
+   public CompletionStage<Void> rollbackAllTxStores(TxInvocationContext<AbstractCacheTransaction> txInvocationContext,
+                                                    Predicate<? super StoreConfiguration> predicate) {
+      return decorateCommand("rollbackAllTxStores", super.rollbackAllTxStores(txInvocationContext, predicate));
+   }
+
+   @Override
+   public <K, V> CompletionStage<Void> writeEntries(Iterable<MarshallableEntry<K, V>> iterable,
+                                                    Predicate<? super StoreConfiguration> predicate) {
+      return decorateCommand("writeEntries", super.writeEntries(iterable, predicate));
+   }
+
+   @Override
+   public CompletionStage<Long> performBatch(TxInvocationContext<AbstractCacheTransaction> ctx,
+                                             TriPredicate<? super WriteCommand, Object, MVCCEntry<?, ?>> commandKeyPredicate) {
+      return decorateCommand("performBatch", super.performBatch(ctx, commandKeyPredicate));
+   }
+
+   private InfinispanSpan<Object> requestStart(String operationName) {
+      InfinispanSpanAttributes attributes = new InfinispanSpanAttributes.Builder(SpanCategory.PERSISTENCE)
+            .withCache(cache.getName(), configuration)
+            .build();
+
+      return telemetryService.startTraceRequest(operationName, attributes);
+   }
+
+   private <T> CompletionStage<T> decorateCommand(String operationName, CompletionStage<T> operation) {
+      if (super.isEnabled() && configuration.tracing().enabled(SpanCategory.PERSISTENCE)) {
+         operation.whenComplete(requestStart(operationName));
```

Comment:
not sure if you noticed, but this is wrong. The span is started after the operation is in progress (or completed).

Reply:
To be honest I'm not 100% really sure about what you said.
Probably I'm wrong, but going on debugging and analyzing the stack traces I think that the behavior is correct.

When OpenTelemetryService#startTraceRequest is invoked the stack trace is:

The thread name is `expiration-thread-TracingPersistenceTest-NodeA-p5-t1`
``` bash
startTraceRequest:31, OpenTelemetryService (org.infinispan.server.core.telemetry)
requestStart:108, TracedPersistenceManager (org.infinispan.persistence.manager)
decorateCommand:113, TracedPersistenceManager (org.infinispan.persistence.manager)
writeToAllNonTxStores:76, TracedPersistenceManager (org.infinispan.persistence.manager)
storeEntry:469, CacheWriterInterceptor (org.infinispan.interceptors.impl)
storeEntry:460, CacheWriterInterceptor (org.infinispan.interceptors.impl)
```

so it seems that we start the trace request exactly when we invoke the TracedPersistenceManager#writeToAllNonTxStores

Conversely, when OpenTelemetrySpan#complete is invoked the stack trace is:
``` bash
complete:29, OpenTelemetrySpan (org.infinispan.server.core.telemetry)
accept:18, InfinispanSpan (org.infinispan.telemetry)
accept:5, InfinispanSpan (org.infinispan.telemetry)
uniWhenComplete:863, CompletableFuture (java.util.concurrent)
uniWhenCompleteStage:887, CompletableFuture (java.util.concurrent)
whenComplete:2357, CompletableFuture (java.util.concurrent)
whenComplete:144, CompletableFuture (java.util.concurrent)
```

That seems correct to me. But as I said, maybe I'm wrong...

Reply:
I resolve the conversation. @pruivo please unresolve it if you don't agree with my analysis.

Reply:
`PersistentManagerImpl.writeToAllNonTxStores` is invoked before `startTraceRequest`.
I'm expecting to see 
```
OpenTelemetryService.startTraceRequest
PersistentManagerImpl.writeToAllNonTxStores
OpenTelemetrySpan.complete (on CompletableFuture)
```

Reply:
With the last push we trace also the synchronous part of the call.